Data supplier, listeners, assertions (fixes #3):

emilshina opened the pull request:
- added quality plugin;
- created data suppliers;
- added listeners;
- created custom annotation;
- added assertions;
- used lombok plugin.

sskorol commented on src/main/java/com/github/emilshina/listeners/TestListener.java:
```diff
@@ -0,0 +1,31 @@
+
+public class TestListener implements IInvokedMethodListener {
+
+    private static final Logger LOG = Logger.getLogger(TestListener.class.getName());
```
try to replace classic logger with `@Slf4j` annotation, present in Lombok

sskorol commented on build.gradle:
```diff
@@ -10,15 +14,25 @@ repositories {
 }
 
 test {
+//    testLogging {
```
it's not a good practice to push commented code; would be better to remove, if it's not used

sskorol commented on src/test/java/com/github/emilshina/testcases/AuthorizationTests.java:
```diff
@@ -1,33 +1,59 @@
     @BeforeMethod
-    public void beforeEachMethod() {
-        System.out.println("Before method of Auth class.");
+    public void beforeEachMethod(ITestContext context, Method method) {
```
missing `final` keyword... try running quality check again

sskorol commented on src/test/java/com/github/emilshina/testcases/DataSuppliers.java:
```diff
@@ -0,0 +1,50 @@
+
+    private static final Logger LOG = Logger.getLogger(DataSuppliers.class.getName());
+
+    @DataSupplier()
```
redundant `()`

emilshina commented:
fixed all comments

sskorol commented on src/main/java/com/github/emilshina/listeners/TestListener.java:
```diff
@@ -0,0 +1,27 @@
+
+    @Override
+    public void beforeInvocation(final IInvokedMethod method, final ITestResult testResult) {
+        if (method.isTestMethod() && log.isInfoEnabled()) {
```
with slf4j logger you don't need to check if corresponding type is enabled

sskorol commented on src/test/java/com/github/emilshina/testcases/DataSuppliers.java:
```diff
@@ -0,0 +1,48 @@
+
+    @DataSupplier
+    public Map<String, String> userCredentials() {
+        final HashMap<String, String> credentials = new HashMap<>();
```
on the left side it's always better to use interface instead of a concrete type

sskorol commented on src/test/java/com/github/emilshina/testcases/AuthorizationTests.java:
```diff
@@ -1,33 +1,56 @@
+        assertThat(user).hasEmails(asList("email10@test.com", "email11@test.com"));
+
+
+/*      final SoftAssertions softAssertions = new SoftAssertions();
```
commented code shouldn't be included into PR

sskorol commented on src/main/java/com/github/emilshina/utils/CsvUtils.java:
```diff
@@ -0,0 +1,40 @@
+
+import static java.util.Arrays.asList;
+
+public class CsvUtils {
```
utility class must be final

sskorol commented on src/main/java/com/github/emilshina/utils/CsvUtils.java:
```diff
@@ -0,0 +1,40 @@
+
+        try (Stream<String> lines = Files.lines(Paths.get(path))) {
+            List<String> values = lines
+                    .filter(line -> line.startsWith("3099"))
```
it doesn't seem to be generic anymore, if you use magic numbers / hardcoded strings for filtering

sskorol commented on src/main/java/com/github/emilshina/utils/CsvUtils.java:
```diff
@@ -0,0 +1,40 @@
+        try (Stream<String> lines = Files.lines(Paths.get(path))) {
+            List<String> values = lines
+                    .filter(line -> line.startsWith("3099"))
+                    .flatMap(line -> (Arrays.stream(line.split(";"))))
```
redundant `()` for wrapping `Arrays`

sskorol commented on src/main/java/com/github/emilshina/utils/CsvUtils.java:
```diff
@@ -0,0 +1,40 @@
+            List<String> values = lines
+                    .filter(line -> line.startsWith("3099"))
+                    .flatMap(line -> (Arrays.stream(line.split(";"))))
+                    .collect(Collectors.toList());
```
why don't you use `StreamEx` with `toList()` terminal operation?

sskorol commented on src/main/java/com/github/emilshina/utils/CsvUtils.java:
```diff
@@ -0,0 +1,40 @@
+                    .flatMap(line -> (Arrays.stream(line.split(";"))))
+                    .collect(Collectors.toList());
+
+            return new User(Long.parseLong(values.get(0)), values.get(1), values.get(2), asList(values.get(3)));
```
a good example of using reflection... you can't init a new object of generic type without reflection... and otherwise, your method won't be generic...

sskorol commented on src/main/java/com/github/emilshina/utils/CsvUtils.java:
```diff
@@ -0,0 +1,40 @@
+            return new User(Long.parseLong(values.get(0)), values.get(1), values.get(2), asList(values.get(3)));
+
+        } catch (IOException e) {
+            e.printStackTrace();
```
it's not recommended to print stacktrace into system out thread... better to use logger... + it seems like you need to run `check` task to analyze your code...

sskorol commented on src/main/java/com/github/emilshina/utils/CsvUtils.java:
```diff
@@ -0,0 +1,40 @@
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+
```
redundant new lines

sskorol commented on src/test/java/com/github/emilshina/testcases/AuthorizationTests.java:
```diff
@@ -39,14 +38,14 @@ public void userShouldBeAuthorizedWithValidCredentials(final HashMap.SimpleEntry
-        softAssertions.assertThat(user.getUserName()).as("Username verification.").isEqualTo("username1");
-        log.info("Invalid credentials for: {}.", user.getUserName());
-        softAssertions.assertAll();*/
+        /*      final SoftAssertions softAssertions = new SoftAssertions();
```
commented code is not a good candidate for merging

sskorol commented on src/test/java/com/github/emilshina/testcases/DataSuppliers.java:
```diff
@@ -19,20 +18,21 @@
 
-        return StreamEx.of(
+
+        /*return StreamEx.of(
```
remove commented code

emilshina commented:
@sskorol Please review the last changes.
- implemented spi;
- created the test based on spi (MenuTests.java);
- added dataType into annotation Data;
- created DataSupplier for read data based on dataType (DataSuppliers.java).

sskorol commented on src/main/java/com/github/emilshina/assertions/UserAssert.java:
```diff
@@ -0,0 +1,52 @@
+
+        final List<String> actualEmails = actual.getEmails();
+
+        if (!Objects.areEqual(actualEmails, expectedEmails)) {
```
for collections it's better to use special `Iterables` instance

sskorol replied:
it's not yet fixed

sskorol commented on src/test/java/com/github/emilshina/testcases/DataSuppliers.java:
```diff
@@ -0,0 +1,44 @@
+        final String source = method.getDeclaredAnnotation(Data.class).source();
+        return StreamEx.of(ServiceLoaderUtils.load(DataReader.class, ClassLoader.getSystemClassLoader()))
+                .findFirst(dataReader -> dataReader.dataType().equals(dataTypeFromAnnotation))
+                .map(dataReader -> dataReader.readFrom(source, User.class))
```
try passing entity class within `Data` annotation and make a generic data supplier

emilshina commented:
@sskorol Please review the latest changes

sskorol commented on src/main/java/com/github/emilshina/data/DataReader.java:
```diff
@@ -0,0 +1,12 @@
+
+public interface DataReader {
+
+    <T> T[] readFrom(String dataSource, Class<T> entityClass, Class<T[]> entityArrClass);
```
you need to revise this logic... user have to specify either single class on array... if we force them putting both, it would look strange from design perspective...

sskorol commented on src/main/java/com/github/emilshina/utils/JsonParserUtils.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <T> T[] fromJsonAsArray(final Gson gson, final JsonElement json, final Class<T> tClass,
```
you need to revise this logic to accept a class which could be either array or a single item... and convert a single item to array if required...

sskorol commented on src/test/java/com/github/emilshina/data/CsvReader.java:
```diff
@@ -0,0 +1,47 @@
+        try (Stream<String> stream = Files.lines(Paths.get(path))) {
+            objects = StreamEx.of(stream)
+                    .skip(1)
+                    .map(line -> line.split(";"))
```
what if a user wants to use another separator?.. e.g. `,`... try to move it into properties to give users choice...

sskorol commented on src/test/java/com/github/emilshina/data/CsvReader.java:
```diff
@@ -0,0 +1,47 @@
+                    .skip(1)
+                    .map(line -> line.split(";"))
+                    .map(data -> getObject(entityClass, (Object[]) data))
+                    .collect(Collectors.toList());
```
why don't you use `toList()` directly?

sskorol commented on src/test/java/com/github/emilshina/data/CsvReader.java:
```diff
@@ -0,0 +1,47 @@
+        return "csv";
+    }
+
+    private <T> T getObject(final Class<T> entityClass, final Object... args) {
```
it doesn't seem to be a good candidate for putting into exact implementation... probably better to move it into utility class?

sskorol commented on src/test/java/com/github/emilshina/data/JsonReader.java:
```diff
@@ -0,0 +1,36 @@
+        final Gson gson = new Gson();
+        final String path = ClassLoader.getSystemResource(dataSource).getPath();
+        try (FileReader fileReader = new FileReader(path)) {
+            final com.google.gson.stream.JsonReader reader = new com.google.gson.stream.JsonReader(fileReader);
```
`JsonReader` seems to be closable as well... move to `try` block

sskorol commented on src/test/java/com/github/emilshina/testcases/Data.java:
```diff
@@ -0,0 +1,22 @@
+
+    Class entity() default Object.class;
+
+    Class entityArr() default Object[].class;
```
it's not required to explicitly set parameters for single class and array... you can use only one of them... generic type is more valuable... + your data processor should be smart enough to determine if there's a single object or array in a file, and make corresponding conversions on fly...

sskorol commented on src/test/java/com/github/emilshina/testcases/Data.java:
```diff
@@ -0,0 +1,22 @@
+
+    String source() default "";
+
+    String dataType() default "";
```
it makes no sense to pass type through annotation, as you already have a file extension and overridden interface method... let's assume that you'll work only with files...

sskorol commented on src/test/java/com/github/emilshina/testcases/MenuTests.java:
```diff
@@ -1,27 +1,44 @@
+        log.info("Contacts page was opened.");
     }
 
+    @Data(source = "data.csv", dataType = "CSV")
```
try to create tests for covering both single and multiple objects reading from files...

emilshina commented:
@sskorol Comments were fixed, please review changes.

sskorol commented on src/test/java/com/github/emilshina/data/CsvReader.java:
```diff
@@ -0,0 +1,36 @@
+        return (T[]) Try.withResources(() -> Files.lines(Paths.get(getSystemResource(dataSource).getPath())))
+                .of(stringStream -> StreamEx.of(stringStream)
+                        .skip(1)
+                        .map(line -> line.split(";"))
```
do you want to move separator into properties?

emilshina replied:
Will do it in another branch, because the file with properties is created in branch 4-webdriver-tests. I want to avoid an additional merge.
